### PR TITLE
Adds CLAHE equalization

### PR DESCRIPTION
**Pull request by lferraz**
### Description
Add the clahe algorithm.

### Status
**Work in progress**

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/enhance/equalization.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+from typing import Tuple
+import math
+
+import torch
+import torch.nn.functional as F
+
+from kornia.enhance.histogram import histogram
+from kornia.utils.image import _to_bchw
+
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:11]: add here `__all__ = ["equalize_clahe"`]

File: kornia/enhance/equalization.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    Args:
+        tiles_x_im (torch.Tensor): set of tiles per image to apply the lut. (B, GH, GW, C, TH, TW)
+        num_bins (int, optional): number of bins. default: 256
+        clip (float): threshold value for contrast limiting. If it is 0 then the clipping is disabled. Default: 40.
+        diff (bool, optional): denote if the differentiable histagram will be used. Default: False
+
+    Returns:
+        torch.Tensor: Lut for each tile (B, GH, GW, C, 256)
+
+    """
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:106]: input assertion shape  ?

File: kornia/enhance/equalization.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        for i, tile in enumerate(tiles.unbind(0)):
+            histos[i] = torch.histc(tile, bins=num_bins, min=0, max=1)
+    else:
+        bins: torch.Tensor = torch.linspace(0, 1, num_bins, device=tiles.device)
+        histos = histogram(tiles, bins, torch.tensor(0.001)).squeeze()
+        histos *= pixels
+
+    # clip limit (TODO: optimice the code)
+    if clip > 0:
+        clip_limit = clip * pixels // num_bins
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:120]: typing

File: kornia/enhance/equalization.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+                hist[idxs] = clip_limit
+
+                redist = clipped // num_bins
+                hist += redist
+
+                residual = clipped - redist * num_bins
+                if residual:
+                    hist[0:int(residual)] += 1
+
+    lut_scale = (num_bins - 1) / pixels
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:137]: typing

File: kornia/enhance/equalization.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    r"""Assign the required luts to each tile.
+
+    Args:
+        interp_tiles (torch.Tensor): set of interpolation tiles. (B, 2GH, 2GW, C, TH/2, TW/2)
+        luts (torch.Tensor): luts for each one of the original tiles. (B, GH, GW, C, 256)
+
+    Returns:
+        torch.Tensor: mapped luts (B, 2GH, 2GW, 4, C, 256)
+
+    """
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:154]: assert inputs shape

File: kornia/enhance/equalization.py [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+
+    return luts_x_interp_tiles
+
+
+def _compute_equalized_tiles(interp_tiles: torch.Tensor, luts: torch.Tensor) -> torch.Tensor:
+    r"""Equalize the tiles.
+
+    Args:
+        interp_tiles (torch.Tensor): set of interpolation tiles. (B, 2GH, 2GW, C, TH/2, TW/2)
+        luts (torch.Tensor): luts for each one of the original tiles. (B, GH, GW, C, 256)
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:196]: is 256 fixed ? can this be dynamic ?

File: kornia/enhance/equalization.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+    return tiles_equalized.to(interp_tiles).div(255.)
+
+
+def equalize_clahe(input: torch.Tensor, clip_limit: float = 40., grid_size: Tuple[int, int] = (8, 8)) -> torch.Tensor:
+    r"""Apply clahe equalization on the input tensor.
+
+    Args:
+        input (torch.Tensor): images tensor to equalize with shapes like :math:`(C, H, W)` or :math:`(B, C, H, W)`.
+        clip_limit (float): threshold value for contrast limiting. If 0 clipping is disabled. Default: 40.
+        grid_size (Tuple[int, int]): number of tiles to be cropped in each direction (GH, GW).
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:262]: add default values

File: kornia/enhance/equalization.py [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+
+    Returns:
+        torch.Tensor: Equalized image or images with shape as the input.
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:265]: add a small example here so that (for doctest sanity check)

File: kornia/enhance/equalization.py [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+
+    """
+    if not torch.is_tensor(input):
+        raise TypeError(f"Input input type is not a torch.Tensor. Got {type(input)}")
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:269]: add end line in between

File: kornia/enhance/equalization.py [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+    if grid_size[0] <= 0 or grid_size[1] <= 0:
+        raise ValueError("Input grid_size elements must be positive. Got {grid_size}")
+
+    imgs: torch.Tensor = _to_bchw(input)  # B x C x H x W
+
+    hist_tiles: torch.Tensor  # B x GH x GW x C x TH x TW
+    img_padded: torch.Tensor  # B x C x H' x W'
+    # the size of the tiles must be even in order to divide them into 4 tiles for the interpolation
+    hist_tiles, img_padded = _compute_tiles(imgs, grid_size, True)
+    tile_size: Tuple[int, int] = hist_tiles.shape[-2:]  # type: ignore
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:291]: why ignore?

File: kornia/enhance/equalization.py [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+
+    # reconstruct the images form the tiles
+    eq_imgs: torch.Tensor = torch.cat(equalized_tiles.unbind(2), 4)
+    eq_imgs = torch.cat(eq_imgs.unbind(1), 2)
+    h, w = imgs.shape[-2:]
+    eq_imgs = eq_imgs[..., :h, :w]  # crop imgs if they were padded
+
+    # remove batch if the input was not in batch form
+    if input.dim() != eq_imgs.dim():
+        eq_imgs.squeeze_(0)
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:305]: inplace stuff usually goes wrong with diff

**Review by edgarriba — CHANGES_REQUESTED** [review]
Minor comments for now - need tests :)

File: kornia/enhance/equalization.py (hunk 6 above)
Comment (lferraz) [reply kornia/enhance/equalization.py:196]: Now is fixed, in the next version will be dynamic.

File: kornia/enhance/equalization.py (hunk 10 above)
Comment (lferraz) [reply kornia/enhance/equalization.py:291]: mypy cannot infer the type. I added a comment.

File: kornia/enhance/equalization.py (hunk 9 above)
Comment (lferraz) [reply kornia/enhance/equalization.py:269]: f"Input input type is not a torch.Tensor.\n Got {type(input)}"?

Comment (edgarriba) [reply kornia/enhance/equalization.py:297]: Between if blocks

File: kornia/enhance/equalization.py [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+    if not torch.is_tensor(input):
+        raise TypeError(f"Input input type is not a torch.Tensor. Got {type(input)}")
+
+    if input.dim() not in [3, 4]:
+        raise ValueError(f"Invalid input shape, we expect CxHxW or BxCxHxW. Got: {input.shape}")
+
+    if input.numel() == 0:
+        raise ValueError("Invalid input tensor, it is empty.")
+
+    if input.max() > 1. or input.min() < 0.:
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:305]: I would remove this check as we dicussed in the augmentations might slow down processes. /cc @shijianjian

**Review by edgarriba — COMMENTED** [review]
implement test_gradcheck/jit and good to go

**Review by edgarriba — APPROVED** [review]
LGTM. One last missing thing - you need to rebase
